File: Artificial Intelligence/Ass No-04 Graph_coloring.py
def add_edge(graph, u, v):
    graph[u][v] = 1
    graph[v][u] = 1

def is_safe(node, color, graph, c, V):
    for i in range(V):
        if graph[node][i] == 1 and color[i] == c:
            return False
    return True

def solve(i, graph, m, V, color):
    if i == V:
        return True

    for c in range(1, m + 1):
        if is_safe(i, color, graph, c, V):
            color[i] = c
            if solve(i + 1, graph, m, V, color):
                return True
            color[i] = 0  # backtrack

    return False

def graph_coloring(graph, m, V):
    color = [0] * V
    if solve(0, graph, m, V, color):
        print("Solution exists.")
        print("Coloring of vertices:", color)
    else:
        print("No solution exists.")

# Main Execution
V = 5  # Number of vertices
m = 3  # Number of colors

# Initialize adjacency matrix
graph = [[0 for _ in range(V)] for _ in range(V)]

# Add edges
add_edge(graph, 0, 1)
add_edge(graph, 0, 2)
add_edge(graph, 0, 3)
add_edge(graph, 2, 4)
add_edge(graph, 3, 4)

# Solve the problem
graph_coloring(graph, m, V)


# Output
# Solution exists.
# Coloring of vertices: [1, 2, 2, 2, 1]

File: Artificial Intelligence/Ass No-04 Graph_coloring.cpp
#include <iostream>
#include <vector>
using namespace std;

void addEdge(int graph[5][5], int edge1, int edge2) {
    graph[edge1][edge2] = 1;
    graph[edge2][edge1] = 1;
}

bool safeToAssign(int i, int j, int graph[5][5], int v, vector<int> &color) {
    for (int k = 0; k < v; k++) {
        if (graph[i][k] == 1 && color[k] == j) {
            return false; // If adjacent node has the same color, return false
        }
    }
    return true;
}

bool function(int graph[5][5], int m, int v, int i, vector<int> &color) {
    if (i == v) {
        return true; // If all vertices are colored successfully
    }

    for (int j = 0; j < m; j++) {
        if (safeToAssign(i, j, graph, v, color)) {
            color[i] = j; // Assign color
            if (function(graph, m, v, i + 1, color)) return true; // Recurse for the next vertex
            color[i] = -1; // Backtrack if coloring doesn't work
        }
    }
    return false;
}

bool graphColoring(int graph[5][5], int m, int v) {
    vector<int> color(v, -1); // Initialize color array with -1 for all vertices
    int i = 0; // Start with the first vertex
    return function(graph, m, v, i, color);
}

int main() {
    int m = 3;  // Number of colors
    int v = 5;  // Number of vertices
    int graph[5][5] = {0};  // Initialize graph with all zeros (no edges)

    // Adding edges
    addEdge(graph, 0, 1);
    addEdge(graph, 0, 2);
    addEdge(graph, 0, 3);
    addEdge(graph, 2, 4);
    addEdge(graph, 3, 4);

    if (graphColoring(graph, m, v)) {
        cout << "Solution exists." << endl;
    } else {
        cout << "No solution exists." << endl;
    }

    return 0;
}

#Output
#Solution exists.
